[Gloo] Support work-level timeouts in ProcessGroupGloo

osalpekar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #41265 [Gloo] Tests for Gloo Async Work Wait-level Timeouts
* **#40948 [Gloo] Support work-level timeouts in ProcessGroupGloo**
* #40947 [NCCL] Tests for WorkNCCL::wait with Timeouts
* #40946 [NCCL] Support Wait Timeout in ProcessGroupNCCL
* #40944 [NCCL] Add timeout to ProcessGroup Work Wait

Add work-level timeouts to ProcessGroupGloo. This uses the timeout support in `waitSend` and `waitRecv` functions from Gloo's `unbound_buffer` construct.

Differential Revision: [D22173763](https://our.internmc.facebook.com/intern/diff/D22173763/)

osalpekar committed:
[Gloo] Support work-level timeouts in ProcessGroupGloo

Add work-level timeouts to ProcessGroupGloo. This uses the timeout support in `waitSend` and `waitRecv` functions from Gloo's `unbound_buffer` construct.

Differential Revision: [D22173763](https://our.internmc.facebook.com/intern/diff/D22173763/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Gloo] Support work-level timeouts in ProcessGroupGloo"


Add work-level timeouts to ProcessGroupGloo. This uses the timeout support in `waitSend` and `waitRecv` functions from Gloo's `unbound_buffer` construct.

Differential Revision: [D22173763](https://our.internmc.facebook.com/intern/diff/D22173763/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Gloo] Support work-level timeouts in ProcessGroupGloo"


Add work-level timeouts to ProcessGroupGloo. This uses the timeout support in `waitSend` and `waitRecv` functions from Gloo's `unbound_buffer` construct.

Differential Revision: [D22173763](https://our.internmc.facebook.com/intern/diff/D22173763/)

[ghstack-poisoned]

mrshenli reviewed (commented):
Shall we add a test for this?

osalpekar committed:
Update on "[Gloo] Support work-level timeouts in ProcessGroupGloo"


Add work-level timeouts to ProcessGroupGloo. This uses the timeout support in `waitSend` and `waitRecv` functions from Gloo's `unbound_buffer` construct.

Differential Revision: [D22173763](https://our.internmc.facebook.com/intern/diff/D22173763/)

[ghstack-poisoned]

osalpekar commented:
> Shall we add a test for this?

Added tests in the next PR in this stack (#41265)

osalpekar committed:
Update on "[Gloo] Support work-level timeouts in ProcessGroupGloo"


Add work-level timeouts to ProcessGroupGloo. This uses the timeout support in `waitSend` and `waitRecv` functions from Gloo's `unbound_buffer` construct.

Differential Revision: [D22173763](https://our.internmc.facebook.com/intern/diff/D22173763/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Gloo] Support work-level timeouts in ProcessGroupGloo"


Add work-level timeouts to ProcessGroupGloo. This uses the timeout support in `waitSend` and `waitRecv` functions from Gloo's `unbound_buffer` construct.

Differential Revision: [D22173763](https://our.internmc.facebook.com/intern/diff/D22173763/)

[ghstack-poisoned]

osalpekar committed:
Update on "[Gloo] Support work-level timeouts in ProcessGroupGloo"


Add work-level timeouts to ProcessGroupGloo. This uses the timeout support in `waitSend` and `waitRecv` functions from Gloo's `unbound_buffer` construct.

Differential Revision: [D22173763](https://our.internmc.facebook.com/intern/diff/D22173763/)

[ghstack-poisoned]
